### PR TITLE
Cope with Pycharm's oddity in detecting namespace packages

### PR DESCRIPTION
**Pull request by ashb**
Without the `__path__` assignment being the very first line of the file
(including before any import) it doesn't detect this correctly.

### CODE REVIEW

**Comment by ashb** [comment]
@shahar1 @kaxil Could one of you test if this fixes it please?

**Comment by kaxil** [comment]
> @shahar1 @kaxil Could one of you test if this fixes it please?

Yup, works now

**Review by kaxil — APPROVED** [review]

File: airflow/__init__.py [hunk 1]
```diff
@@ -15,13 +15,15 @@
...
 # under the License.
-from __future__ import annotations
 
 # Make `airflow` a namespace package, supporting installing
 # airflow.providers.* in different locations (i.e. one in site, and one in user
 # lib.)  This is required by some IDEs to resolve the import paths.
+# And it _has_ to be the first code line too. Sad panda
 __path__ = __import__("pkgutil").extend_path(__path__, __name__)  # type: ignore
 
+from __future__ import annotations  # noqa: F404
```
Comment (kaxil) [inline airflow/__init__.py:25]: Although:

>Python mandates that future-imports must appear in the module before any other code except docstrings

https://peps.python.org/pep-0008/#module-level-dunder-names

File: providers/src/airflow/providers/__init__.py [hunk 2]
```diff
@@ -16,11 +16,13 @@
 # specific language governing permissions and limitations
 # under the License.
 
-from __future__ import annotations
 
 # Make `airflow` a namespace package, supporting installing
 # airflow.providers.* in different locations (i.e. one in site, and one in user
 # lib.)  This is required by some IDEs to resolve the import paths.
+# And it _has_ to be the first code line too. Sad panda
```
Comment (shahar1) [inline providers/src/airflow/providers/__init__.py:23]: 🐼☹

File: airflow/__init__.py (hunk 1 above)
Comment (shahar1) [reply airflow/__init__.py:25]: Question to the audience - can we live with it (or similar workaround) for now until we find a better alternative?

Partial answer - mypy and some static checks don't like it...

Comment (ashb) [reply airflow/__init__.py:25]: This really does feel like a massive bug in pycharm :(

**Comment by vincbeck** [comment]
Yeah I tried in #43081 but could not make it work. Mypy/Python is not happy about it and raise a `SyntaxError`

File: airflow/__init__.py (hunk 1 above)
Comment (kaxil) [reply airflow/__init__.py:25]: Yeah, for now I think we can live with that bug.

Once we move a good number of things from Airflow to SDK, maybe the `airflow/__init__.py` file won't have much content left to keep `from __future__ import annotations`!

**Comment by ashb** [comment]
https://youtrack.jetbrains.com/issue/PY-46690/Allow-all-imports-from-the-same-named-packages-if-path-manipulation-in-init-is-detected is the closest current issue I've found, but not exactly the same

**Comment by potiuk** [comment]
Let's just exclude `airflow.__init__.py` from automatically adding `from __future__ import annotatios` and convert it back to use "regular" typing.

We've done that for Pydantic classes before and still doing it for some test files https://github.com/apache/airflow/blob/main/pyproject.toml#L355

**Comment by potiuk** [comment]
I added alternative implementation - and it seems to work for me https://github.com/apache/airflow/pull/43173 - just by removing `from future` and adding ignore rules from ruff seem to do the job as I suspected. Actually `airflow/__init__.py` did not require any change beceause it does not use postponed evaluation of types PEP563 at all.

**Comment by potiuk** [comment]
Closing this one. It's already fixed.

**Pull request closed without merging**